Python/Module2_EssentialsOfPython/Errors_and_Debugging.md

```diff
@@ -0,0 +1,112 @@
+
+```python
+# attempting to take the square root of a negative number
+>>> from math import sqrt
+>>> math.sqrt(-1)
+ValueError: math domain error
+
+```
+
+The math library cannot take the square root of a negative number. Because the number passed as an argument is technically the correct type, but is not compatible with the function, a `ValueError` is raised. Whenever you encounter one of these exceptions, be sure that you understand why the encounter is raised before you try to fix it.
+
+It is also possible to raise exceptions on purpose with the `raise` statement. If you are writing a function that is designed only for certain inputs, it can be worthwhile to verify that inputs will be valid. As an example, let's create our own square root function.
+
+```python
+def squareRoot(x):
+  if x < 0:
+    raise ValueError('Cannot take the square root of a negative number.')
+  return x**0.5
+  
+>>> print(squareRoot(-1))
+ValueError: Cannot take the square root of a negative number.
+```
+
+When the invalid value is passed, our function raises a ValueError rather than allowing the function to run with the value, possibly causing a worse or more confusing error. Clearly, in this case it was much better to raise an exception than to allow the program to continue! The `assert` statement can also be used to raise errors in the format `assert [boolean statement], 'error message'`, but it tends to be less informative and less powerful than simply using `raise`.
+
+It is also possible to create custom exceptions using classes, which will be discussed further in Module 4.
+<!-- #endregion -->
+
+
+<!-- #region -->
+## Logic Errors
+
+Logic errors occur when your code will run in the interpreter, but does not behave as expected. Whenever you write a function that works but outputs something incorrect, you have committed a logic error. Logic errors are very easy to commit and can be incredibly difficult to solve. It's impossible to completely avoid logic errors, but if you take care while programming and debugging, they can often be avoided or fixed.
+
+Before writing a complex program, it is often smart to write psuedocode beforehand to ensure that your algorithm makes sense and will do what you want it to. It is important to understand how your program will work before you write it. After writing the program, it can be helpful to take another walk through how the program works, and make sure that every step makes sense. Some programmers keep rubber ducks at their desks and try to explain their code to the duck when they hit a bug. Sometimes, simply walking through your code slowly and describing everything can help you see an error that you otherwise may have missed.
```

Comment:
Pseudocode isn't mentioned anywhere in the site as far as I can tell. Where do you think it would be good to include a discussion on pseudocode? I don't want to overcomplicate this section but it's definitely something we should talk about